jib-core/src/main/java/com/google/cloud/tools/jib/builder/AsyncStep.java

```diff
@@ -25,21 +25,19 @@
  * A step to run asynchronously. Implementations should:
  *
  * <ol>
- *   <li>Construct with the dependent {@link AsyncStep}s.
+ *   <li>Be immutable
+ *   <li>Construct with the dependent {@link AsyncStep}s and submitting itself to the {@link
+ *       ListeningExecutorService} to run after all its dependent {@link AsyncStep}s (for example,
+ *       by using {@link Futures#whenAllSucceed})
  *   <li>Implement {@link #call} with the actual work
- *   <li>Implement {@link #getFuture} by submitting itself to the {@link ListeningExecutorService}
- *       to run after all its dependent {@link AsyncStep}s for the first call, and returning that
- *       future for subsequent calls.
+ *   <li>Implement {@link #getFuture} by
  * </ol>
  *
  * @param <T> the object type passed on by this step
  */
 interface AsyncStep<T> extends Callable<T> {
 
-  /**
-   * Submits to a {@link ListeningExecutorService} to run after its dependencies. For example, by
-   * using {@link Futures#whenAllSucceed}.
-   */
+  /** @return the submitted future */
```

Comment:
Is anyone calling this based on the interface type? Or are they just all calling this concrete instance of the step object? In that case, is this interface needed?

Reply:
Yeah, there is for example, `PushBlobStep`, that uses an `AsyncStep<CachedLayer>` to be either base image or application layers.